src/JustSaying/Messaging/Channels/SubscriptionGroups/SubscriptionGroupFactory.cs

```diff
@@ -0,0 +1,106 @@
+{
+    public class SubscriptionGroupFactory : ISubscriptionGroupFactory
+    {
+        private readonly IMessageDispatcher _messageDispatcher;
+        private readonly IMessageMonitor _monitor;
+        private readonly ILoggerFactory _loggerFactory;
+
+        public SubscriptionGroupFactory(
+            IMessageDispatcher messageDispatcher,
+            IMessageMonitor monitor,
+            ILoggerFactory loggerFactory)
+        {
+            _messageDispatcher = messageDispatcher;
+            _monitor = monitor;
+            _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        public SubscriptionGroupCollection Create(
+            SubscriptionConfigBuilder defaults,
+            IDictionary<string, SubscriptionGroupConfigBuilder> consumerGroupSettings)
+        {
+            List<ISubscriptionGroup> groups = consumerGroupSettings
+                .Values
+                .Select(builder => Create(defaults, builder.Build(defaults)))
+                .ToList();
+
+            return new SubscriptionGroupCollection(
+                groups,
+                _loggerFactory.CreateLogger<SubscriptionGroupCollection>());
+        }
+
+        private ISubscriptionGroup Create(SubscriptionConfigBuilder defaults, SubscriptionGroupSettings settings)
+        {
+            IMultiplexer multiplexer = CreateMultiplexer(settings.MultiplexerCapacity);
+            ICollection<IMessageReceiveBuffer> receiveBuffers = CreateBuffers(defaults, settings);
+            ICollection<IMultiplexerSubscriber> subscribers = CreateSubscribers(settings.ConcurrencyLimit);
+
+            foreach (IMessageReceiveBuffer receiveBuffer in receiveBuffers)
+            {
+                multiplexer.ReadFrom(receiveBuffer.Reader);
+            }
+
+            foreach (IMultiplexerSubscriber subscriber in subscribers)
+            {
+                subscriber.Subscribe(multiplexer.GetMessagesAsync());
+            }
+
+            return new SubscriptionGroup(
+                settings,
+                receiveBuffers,
+                multiplexer,
+                subscribers,
+                _loggerFactory.CreateLogger<SubscriptionGroup>());
+        }
+
+        private ICollection<IMessageReceiveBuffer> CreateBuffers(
+            SubscriptionConfigBuilder defaults,
+            SubscriptionGroupSettings subscriptionGroupSettings)
+        {
+            var buffers = new List<IMessageReceiveBuffer>();
+
+            foreach (ISqsQueue queue in subscriptionGroupSettings.Queues)
+            {
+                var buffer = new MessageReceiveBuffer(
+                    subscriptionGroupSettings.Prefetch,
+                    subscriptionGroupSettings.BufferSize,
+                    subscriptionGroupSettings.ReceiveBufferReadTimeout,
+                    queue,
+                    defaults.SqsMiddleware ?? new DefaultSqsMiddleware(_loggerFactory.CreateLogger<DefaultSqsMiddleware>()),
```

Comment:
Does each queue need its own `DefaultSqsMiddleware`, or could you just create one (if needed) outside the loop?

Reply:
They could probably share, but I see no harm in having one per queue.

Reply:
If there's no shared state, then it would be more efficient to share it to reduce allocations.

Reply:
I've moved this creation to the constructor, so we're just using the same default middleware if one hasn't been specified on the defaults.